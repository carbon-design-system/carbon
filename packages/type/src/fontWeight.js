/**
 * Copyright IBM Corp. 2018, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export const fontWeights = {
  light: 300,
  regular: 400,
  semibold: 600,
};

export function fontWeight(weight) {
  if (!fontWeights[weight]) {
    throw new Error(
      `Unable to find font weight: \`${weight}\`. Expected one of: ` +
        `[${Object.keys(fontWeights).join(', ')}]`
    );
  }
  return {
    fontWeight: fontWeights[weight],
  };
}
