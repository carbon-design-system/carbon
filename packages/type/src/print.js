/**
 * Copyright IBM Corp. 2018, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export function print(block) {
  return Object.keys(block).reduce((acc, key, index) => {
    // Short-circuit on the foreign key 'breakpoints'. This is used in our
    // tokens for fluid type and should not be printed. In the future, we should
    // tie this to media query outputs.
    if (key === 'breakpoints') {
      return acc;
    }

    const property = `${paramCase(key)}: ${block[key]};`;
    if (index === 0) {
      return property;
    }
    return acc + '\n' + property;
  }, '');
}

function paramCase(string) {
  let result = '';
  for (let i = 0; i < string.length; i++) {
    const character = string[i];
    if (character === character.toUpperCase()) {
      result += '-' + character.toLowerCase();
      continue;
    }
    result += character;
  }
  return result;
}
