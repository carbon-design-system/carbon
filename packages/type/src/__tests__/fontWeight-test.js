/**
 * Copyright IBM Corp. 2018, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @jest-environment node
 */

import { fontWeights, fontWeight } from '../fontWeight';
import { print } from '../print';

describe('fontWeight', () => {
  it('should export the supported font weights', () => {
    expect(fontWeights).toMatchSnapshot();
  });

  it('should support getting the quoted string for a font weight', () => {
    expect(fontWeight('light')).toEqual({
      fontWeight: fontWeights.light,
    });
  });

  it('should error out if trying to get a font weight that does not exist', () => {
    expect(() => {
      fontWeight('<unknown>');
    }).toThrow();
  });

  it('should be printable', () => {
    expect(print(fontWeight('regular'))).toMatchSnapshot();
  });
});
