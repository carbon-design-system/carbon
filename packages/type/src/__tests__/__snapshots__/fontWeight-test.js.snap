// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fontWeight should be printable 1`] = `"font-weight: 400;"`;

exports[`fontWeight should export the supported font weights 1`] = `
{
  "light": 300,
  "regular": 400,
  "semibold": 600,
}
`;
